### PR TITLE
[common/containers] feature: add conn pool

### PR DESCRIPTION
**Pull request by drmingdrmer**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

##### [common/containers] feature: add conn pool
- common_containers::Pool caches Channel and indexes them by addresses.

- Raft Network use Pool to reduce number of connections

- Add tracing logs address the time spent by metasrv tests.

- Increase thread number for metasrv tests.

## Changelog

- New Feature

## Related Issues

### CODE REVIEW

File: common/meta/raft-store/src/state_machine/sm.rs [hunk 1]
```diff
@@ -247,12 +247,14 @@ impl StateMachine {
...
     /// will be made and the previous resp is returned. In this way a client is able to re-send a
     /// command safely in case of network failure etc.
-    #[tracing::instrument(level = "trace", skip(self))]
+    #[tracing::instrument(level = "debug", skip(self))]
     pub async fn apply(&self, entry: &Entry<LogEntry>) -> common_exception::Result<AppliedState> {
         // TODO(xp): all update need to be done in a tx.
 
         let log_id = &entry.log_id;
 
+        tracing::debug!("ddd_tx start: {:?}", entry);
```
Comment (ariesdevil) [inline common/meta/raft-store/src/state_machine/sm.rs:256]: The debug should be more formal and readable

File: common/meta/raft-store/src/state_machine/sm.rs [hunk 2]
```diff
@@ -292,6 +294,8 @@ impl StateMachine {
             Ok(None)
         })?;
 
+        tracing::debug!("ddd_tx stop: {:?}", entry);
```
Comment (ariesdevil) [inline common/meta/raft-store/src/state_machine/sm.rs:297]: Same as above

File: common/meta/sled-store/src/sled_tree.rs [hunk 3]
```diff
@@ -94,6 +94,8 @@ impl SledTree {
         sync: bool,
         f: impl Fn(TransactionSledTree<'_>) -> Result<T, UnabortableTransactionError>,
     ) -> Result<T, ErrorCode> {
+        let sync = sync && self.sync;
```
Comment (ariesdevil) [inline common/meta/sled-store/src/sled_tree.rs:97]: Save my ass :)

File: metasrv/src/network.rs [hunk 4]
```diff
@@ -22,30 +23,61 @@ use async_raft::raft::InstallSnapshotResponse;
...
+    conn_pool: Pool<ChannelManager>,
 }
 
 impl Network {
     pub fn new(sto: Arc<MetaRaftStore>) -> Network {
-        Network { sto }
+        let mgr = ChannelManager {};
+        Network {
+            sto,
+            conn_pool: Pool::new(mgr, Duration::from_millis(50)),
```
Comment (ariesdevil) [inline metasrv/src/network.rs:68]: remove the local variable and just pass `ChannelManager{}` to new here?

File: common/containers/src/pool.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                return Ok(itm);
+            } else {
+                // mark broken conn as deleted
+                *guard = None;
+            }
+        }
+
+        let mut interval = self.initial_retry_interval;
+
+        let n_retry = 3;
```
Comment (ariesdevil) [inline common/containers/src/pool.rs:126]: I prefer `n_retry` as a parameter with default value 3

File: metasrv/src/network.rs (hunk 4 above)
Comment (drmingdrmer) [reply metasrv/src/network.rs:68]: 🤔 I'd hate to see more than one things done in one line...

File: common/containers/src/pool.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            manager,
+            err_type: Default::default(),
+        }
+    }
+
+    /// Return an raw pool item.
+    ///
+    /// The returned one may be an uninitialized one, i.e., it contains a None.
+    /// The lock for `items` should not be held for long, e.g. when `build()` a new connection, it takes dozen ms.
+    fn get_pool_item(&self, key: &Mgr::Key) -> PoolItem<Mgr::Item> {
```
Comment (BohuTANG) [inline common/containers/src/pool.rs:90]: 👍 
It would be nice to add eviction for the pool manager : thinking

Comment (drmingdrmer) [reply common/containers/src/pool.rs:90]: Right~

File: common/containers/src/pool.rs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                        return Err(err);
+                    }
+                }
+            }
+
+            sleep(interval).await;
+            interval *= 2;
+        }
+
+        panic!("unreachable")
```
Comment (sundy-li) [inline common/containers/src/pool.rs:151]: unreachable!()

Comment (drmingdrmer) [reply common/containers/src/pool.rs:151]: nice~

**Comment by drmingdrmer** [comment]
2 pending task are tracked in new issues: retry times customization and eviction by LRU.

Time to merge!

**Pull request merged**